File: app/src/main/java/com/example/roombooking/fragments/QRCodeScannerFragment.java
package com.example.roombooking.fragments;

import android.os.Bundle;
import android.util.SparseArray;
import android.view.SurfaceHolder;
import android.view.SurfaceView;
import android.view.View;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.example.roombooking.R;
import com.example.roombooking.utils.CommonUtils;
import com.example.roombooking.utils.ConstKeyBundle;
import com.google.android.gms.vision.CameraSource;
import com.google.android.gms.vision.Detector;
import com.google.android.gms.vision.barcode.Barcode;
import com.google.android.gms.vision.barcode.BarcodeDetector;

import java.io.IOException;

public class QRCodeScannerFragment extends BaseFragment
{
	private SurfaceView 		surfaceView;
	private BarcodeDetector barcodeDetector;
	private CameraSource 	cameraSource;

	private TextView 			txtBarcodeValue;

	@Override
	protected int getLayoutContentID()
	{
		return R.layout.fragment_qrcode_scanner;
	}

	@Override
	public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState)
	{
		super.onViewCreated(view, savedInstanceState);

		surfaceView = view.findViewById(R.id.surface_view);
		txtBarcodeValue = view.findViewById(R.id.txt_barcode_value);
		barcodeDetector = new BarcodeDetector.Builder(getActivity())
				.setBarcodeFormats(Barcode.QR_CODE	)
				.build();
	}

	private void init()
	{
		cameraSource = new CameraSource.Builder(getActivity(), barcodeDetector)
				.setAutoFocusEnabled(true)
				.build();

		surfaceView.getHolder().addCallback(new SurfaceHolder.Callback()
		{
			@Override
			public void surfaceCreated(SurfaceHolder holder)
			{
				try
				{
					cameraSource.start(surfaceView.getHolder());
				}
				catch (IOException e)
				{
					e.printStackTrace();
				}
			}

			@Override
			public void surfaceChanged(SurfaceHolder holder, int format, int width, int height)
			{
			}

			@Override
			public void surfaceDestroyed(SurfaceHolder holder)
			{
				cameraSource.stop();
			}
		});

		barcodeDetector.setProcessor(new Detector.Processor<Barcode>()
		{
			@Override
			public void release()
			{
				//do nothings
			}

			@Override
			public void receiveDetections(Detector.Detections<Barcode> detections)
			{
				final SparseArray<Barcode> barcodes = detections.getDetectedItems();
				if (barcodes.size() != 0)
				{
					final String qrString = barcodes.valueAt(0).rawValue;
					txtBarcodeValue.post(new Runnable()
					{
						@Override
						public void run()
						{
							if(CommonUtils.isQRCodeValid(qrString))
							{
								bookRoom(qrString);
							}
							else
							{
								txtBarcodeValue.setText(getActivity().getResources().getString(R.string.txt_no_valid_code));
							}
						}
					});
				}
				else
				{
					txtBarcodeValue.post(new Runnable()
					{
						@Override
						public void run()
						{
							txtBarcodeValue.setText(getActivity().getResources().getString(R.string.txt_no_valid_code));
						}
					});
				}
			}
		});
	}

	private void bookRoom(String roomName)
	{
		Bundle dataBundle = new Bundle();
		dataBundle.putString(ConstKeyBundle.KEY_ROOM_NAME, roomName);
		switchFragment(new BookingFragment(), dataBundle);
	}

	@Override
	public void onPause()
	{
		super.onPause();
		cameraSource.release();
	}

	@Override
	public void onResume()
	{
		super.onResume();
		init();
	}
}

File: app/src/main/java/com/example/roombooking/fragments/QRCodeScannerFragment.kt
package com.example.roombooking.fragments

import android.os.Bundle
import android.view.SurfaceHolder
import android.view.View
import com.example.roombooking.R
import com.example.roombooking.utils.CommonUtils.isQRCodeValid
import com.example.roombooking.utils.ConstKeyBundle
import com.google.android.gms.vision.CameraSource
import com.google.android.gms.vision.Detector
import com.google.android.gms.vision.Detector.Detections
import com.google.android.gms.vision.barcode.Barcode
import com.google.android.gms.vision.barcode.BarcodeDetector
import kotlinx.android.synthetic.main.fragment_qrcode_scanner.*

class QRCodeScannerFragment : BaseFragment() {
    override val layoutContentID: Int = R.layout.fragment_qrcode_scanner

    private lateinit var barcodeDetector: BarcodeDetector
    private lateinit var cameraSource: CameraSource

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        barcodeDetector = BarcodeDetector.Builder(activity)
                .setBarcodeFormats(Barcode.QR_CODE)
                .build()
    }

    private fun init() {
        cameraSource = CameraSource.Builder(activity, barcodeDetector)
                .setAutoFocusEnabled(true)
                .build()
        surfaceView!!.holder.addCallback(object : SurfaceHolder.Callback {
            override fun surfaceCreated(holder: SurfaceHolder) {
                cameraSource.start(surfaceView!!.holder)
            }

            override fun surfaceChanged(holder: SurfaceHolder, format: Int, width: Int, height: Int) {}
            override fun surfaceDestroyed(holder: SurfaceHolder) {
                cameraSource.stop()
            }
        })
        barcodeDetector.setProcessor(object : Detector.Processor<Barcode> {
            override fun release() {
                //do nothings
            }

            override fun receiveDetections(detections: Detections<Barcode>) {
                val listBarcode = detections.detectedItems
                if (listBarcode.size() != 0) {
                    val qrString = listBarcode.valueAt(0).rawValue
                    textBarcodeValue.post {
                        if (isQRCodeValid(qrString)) {
                            bookRoom(qrString)
                        } else {
                            textBarcodeValue.text = activity!!.resources.getString(R.string.txt_no_valid_code)
                        }
                    }
                } else {
                    textBarcodeValue.post { textBarcodeValue.text = activity!!.resources.getString(R.string.txt_no_valid_code) }
                }
            }
        })
    }

    private fun bookRoom(roomName: String) {
        val dataBundle = Bundle()
        dataBundle.putString(ConstKeyBundle.KEY_ROOM_NAME, roomName)
        switchFragment(BookingFragment(), dataBundle)
    }

    override fun onPause() {
        super.onPause()
        cameraSource.release()
    }

    override fun onResume() {
        super.onResume()
        init()
    }
}